### DISCUSSION TITLE
Deprecated functions in PHTML files

### DISCUSSION THREAD

**Discussion by addison74** [OpenMage/magento-lts#2606] [Questions & Answers]
I recently worked at a store that needs to switch to PHP 8.1. If regarding the OpenMage files the matter was resolved and I did not encounter any errors in the log files, the theme and the extensions caused me big issues. 

Today I spent an hour trying to understand why the product page does not load in the information area. After disabling all the extensions and not getting any positive results, I took the layout files first. I found that removing one loaded the page. I looked inside and identified the few PHTML files. Analyzing them one by one, I found nothing wrong, the variables were all displayed. 

Then I went to the comment line by line until I found ```echo implode ($config, ',')```. When I reversed the parameters of the function, the situation was resolved. I looked for references and starting with PHP 8.0 this function parameters were switched. Honestly, I didn't know about this change, but I can't help but notice that in version 7, the PHP team scratched their right ear with their left hand in some respects.

In case someone faces this problem I wanted to make it public. Please note no errors were reported in the log files. Is there any way to debug the PHTML files and get in system.log the errors?

**Comment by sreichel** [comment]
Have you tried syntax-check that is used in github workflow?

**Reply by addison74** [reply]
I didn't do that. I would like to check the whole theme for syntax. I appreciate if you can give me a recommendation where to start.

**Reply by sreichel** [reply]
GitHub workflow runs this to check syntax errors ... i dont know if your issue is covered

```
- name: Check .phtml files
  run: '! find app/design -type f -name "*.phtml" -exec php -d error_reporting=32767 -l {} \; 2>&1 1> /dev/null | grep "^"'
```

**Reply by sreichel** [reply]
Maybe comment the lines mentioned here helps too ... https://github.com/OpenMage/magento-lts/pull/2586#issue-1368663108

**Comment by kiatng** [comment]
Ref my comment [here](https://github.com/OpenMage/magento-lts/pull/1442#issuecomment-1219044939), can you try to change line 263 to ` } catch (Throwable $e) { ` in file https://github.com/OpenMage/magento-lts/blob/5de30bfc6239d9a4072b743b2dac29c079763176/app/code/core/Mage/Core/Block/Template.php#L263-L266
Does it throw an error on `implode` after the change?

**Reply by addison74** [reply]
Thank you, I forgot about that change. Here is what I get when loading the product page:

```
Fatal error: Uncaught TypeError: implode(): Argument #2 ($array) must be of type ?array, string given in
```

It doesn't appear in exceptions.log file, just in the browser window. That PR must be merged in OpenMage, in both versions.

**Reply by addison74** [reply]
**Moved from https://github.com/OpenMage/magento-lts/pull/1442**

I have a question, is it possible to display these error messages in the exceptions.log file instead of the browser window? If the error occurs in production, at least it is not displayed to visitors.

**Reply by addison74** [reply]
**Moved from #1442** 

@kiatng replied: @ADDISON74 To not display errors in browser, try removing $_SERVER['MAGE_IS_DEVELOPER_MODE'] in your vhost config, and set up the local.xml file in root folder /errors/. The errors will be saved in the folder /var/report/. See also https://magento.stackexchange.com/questions/49355/prevent-error-log-display-to-the-public

**Reply by addison74** [reply]
@kiatng -  I loaded a phpinfo file and I don't have that variable set somewhere. Neither in vhost nor in .htaccess. It is a test DDEV environment, possibly the issue comes from its default settings. I will check more in detail.

**Reply by addison74** [reply]
I found the issue. It comes from ```display_errors = On``` set by default in DDEV. I had to create a new file ```/.ddev/php/noerrors.ini``` and set the value to Off. Once I did this the page is loaded with error 503 and I catch the error message in the log file of the webserver.

**Reply by sreichel** [reply]
@ADDISON74 this might be useful for #2575.

**Comment by addison74** [comment]
At this moment we have the following code in OpenMage: 

https://github.com/OpenMage/magento-lts/blob/f96f0b30ba351d4d6a2a0a0494de6f90c3f85f81/app/code/core/Mage/Core/Block/Template.php#L262-L265

I would propose the following change:

```php
        } catch (Throwable $e) {
            ob_get_clean();
            Mage::logException($e);
            // throw $e;
```

By using ```Mage::logException``` method the error is written in **exceptions.log** file. I commented throw so that the error does not appear in the Frontend anymore. 

The advantage of this simple modification is that the Frontend is loaded entirely and only the block with the issue is omitted. In my case, where the implode function was used, it is not displayed. Now troubleshooting is simple, I go to /var/log and I get all the template issues in exceptions file. 

I'm really surprised why it was preferred to display errors from templates in the browser instead of being recorded in exceptions.log or another file by using ```Mage::log($e, null, 'template_error.log'```.

**Reply by kiatng** [reply]
How about

```php
        } catch (Throwable $e) {
            ob_get_clean();
            Mage::logException($e);
            if (Mage::getIsDeveloperMode()) {
                throw $e;
            }
```

This way, it's more helpful during development.

**Reply by addison74** [reply]
It is a good observation related to the developer mode. Since the error is displayed in the browser, it is no longer needed to write it in the exceptions.log file too. The code would look like this:

```php
} catch (Throwable $e) {
    ob_get_clean();
    if (Mage::getIsDeveloperMode()) {
        throw $e;
    } else {
        Mage::logException($e);
    }
```

**Comment by elidrissidev** [comment]
I have a pending PR that resolves a lot of deprecation errors, but it's not exhaustive. Please if you can, take note of any errors you encounter that have to do with PHP 8.1's changed behavior so that they can be worked on either by me, or anyone else.
